src/unix/tty.c

```diff
@@ -335,6 +335,37 @@ int uv_tty_set_mode(uv_tty_t* tty, uv_tty_mode_t mode) {
 }
 
 
+void uv__tty_close(uv_tty_t* handle) {
+  int expected;
+  int fd;
+
+  fd = handle->io_watcher.fd;
+  if (fd == -1)
+    goto done;
+
+  /* This is used for uv_tty_reset_mode() */
+  do
+    expected = 0;
+  while (!atomic_compare_exchange_strong(&termios_spinlock, &expected, 1));
+
+  if (fd == orig_termios_fd) {
+    /* XXX(bnoordhuis) the tcsetattr is probably wrong when there are still
+     * other uv_tty_t handles active that refer to the same tty/pty but it's
+     * hard to recognize that particular situation without maintaining some
+     * kind of process-global data structure, and that still won't work in a
+     * multi-process setup.
+     */
+    uv__tcsetattr(fd, TCSANOW, &orig_termios);
```

Comment:
This is sort of pre-existing, but why do we use TCSANOW for this? The documentation just says "use TCSADRAIN" if it may be writable. It seems this has some backwards-in-time behavior that may be best to avoid where it changes the way the tty interprets previously written data, but maybe also avoids a stall while the data gets pushed out the serial port. 
```suggestion
    uv__tcsetattr(fd, TCSADRAIN, &orig_termios);
```

Reply:
So, the reason uv_tty_reset_mode() uses TCSANOW is that it can be called from signal handlers. TCSADRAIN and TCSAFLUSH have the potential to block indefinitely, which is undesirable, doubly so inside a signal handler.

(They're also interruptible, meaning you have to have an EINTR retry loop. Also undesirable inside a signal handler.)

That rationale doesn't really apply here so I could change it TCSADRAIN. Alternatively, I could insert a `tcflush(TCIOFLUSH)` before the tcsetattr call. I guess that's a good idea anyway inside uv_tty_reset_mode().

Reply:
I'm going to leave this unchanged for now but plan on revisiting in the future. I noticed that the `tcsetattr(TCSADRAIN)` call in `uv_tty_set_mode()` isn't retried.

(Calling `uv_tty_reset_mode()` afterwards does the right thing, though.)